File: test/magpie/src/bindings/mongo.ts
import { Binding, BindingStatus} from '../binding'
import { MongoClient } from 'mongodb'

// Use this with a value like BINDING_MONGODB_CONNECTIONSTRING
export class MongoBinding implements Binding {
    private uri: string;
    private connection: Promise<MongoClient> | null;

    constructor(map: { [key: string]: string }) {
        this.uri = map['CONNECTIONSTRING'];
        if (!this.uri) {
            throw new Error('CONNECTIONSTRING is required');
        }
    }

    private async connect(): MongoClient {
        if (!this.connection) {
            this.connection = (async () : Promise<MongoClient> => {
                let client = new MongoClient(this.uri);
                await client.connect();
                return client;
            })();
        }

        return this.connection
    }

    public async status(): Promise<BindingStatus> {
        // nothing complex, we just connect.
        await this.connect();
        return { ok: true, message: "connected"};
    }

    public toString = () : string => {
        return 'MongoDB';
    }
}

File: test/magpiego/bindings/mongo.go
package bindings

import (
	"context"
	"log"

	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

// MongoBinding checks if the CONNECTIONSTRING environment parameter is present and if so, attempts to connect to a MongoDB
//
//	instance using the provided URI, returning a BindingStatus indicating success or failure.
func MongoBinding(envParams map[string]string) BindingStatus {
	uri := envParams["CONNECTIONSTRING"]
	if uri == "" {
		log.Println("CONNECTIONSTRING is required")
		return BindingStatus{false, "CONNECTIONSTRING is required"}
	}
	clientOptions := options.Client().ApplyURI(uri)
	ctx := context.Background()
	client, err := mongo.Connect(ctx, clientOptions)
	if err != nil {
		log.Println("mongodb connection failed", err.Error())
		return BindingStatus{false, "mongodb connection failed"}
	}
	err = client.Ping(ctx, nil)
	if err != nil {
		log.Println("mongodb connection failed", err.Error())
		return BindingStatus{false, "mongodb connection failed"}
	}

	return BindingStatus{true, "connected"}
}
